### PR TITLE
assets: add `assets.analyze` and normalize filename behavior of assets plugins

### PR DESCRIPTION
**Pull request by purerosefallen**
`assets.analyze` returns the following structure, which is convenient for implementation of assets plugins.
```ts
  export interface AnalyzeResult {
    buffer: Buffer
    hash: string
    filename: string
  }
```

Also updated all 3 *except remote* assets plugins to use this one.

### CODE REVIEW

File: packages/koishi/src/assets.ts [hunk 1]
```diff
@@ -1,5 +1,8 @@
 import { Context, Service } from '@koishijs/core'
 import { segment } from '@koishijs/utils'
+import { createHash } from 'crypto'
+import { fileTypeFromBuffer } from 'file-type'
```
Comment (Shigma) [inline packages/koishi/src/assets.ts:4]: 这个库导出了一个方法 fromBuffer，或者你直接用默认导出 .fromBuffer 也行

File: packages/koishi/src/assets.ts [hunk 2]
```diff
@@ -27,13 +30,41 @@ export abstract class Assets extends Service {
     const data = await this.ctx.http.get.arraybuffer(url)
     return Buffer.from(data)
   }
+
+  protected async analyze(url: string, file?: string): Promise<Assets.AnalyzeResult> {
+    const buffer = await this.download(url)
+    const hash = createHash('sha1').update(buffer).digest('hex')
+    let filename: string
+    if (file) {
+      const fileBasename = basename(file)
```
Comment (Shigma) [inline packages/koishi/src/assets.ts:39]: 你可以直接 file = basename(file) 看起来更好一点

这个 fileBasename 有点命名方式不统一的味道（camelcase 跟 lowercase 混用了）

**Review by Shigma — COMMENTED** [review]
两个小问题，修了就可以合并

**Comment by Shigma** [comment]
顺便好了以后也可以去 smms 发个 PR。

https://github.com/koishijs/koishi-plugin-assets-smms

File: packages/koishi/src/assets.ts (hunk 1 above)
Comment (purerosefallen) [reply packages/koishi/src/assets.ts:4]: According to [the doc of `file-type`](https://www.npmjs.com/package/file-type), that method does not exist. The `fileTypeFromBuffer` fits our usage here.

File: packages/koishi/src/assets.ts [hunk 3]
```diff
@@ -27,13 +30,42 @@ export abstract class Assets extends Service {
     const data = await this.ctx.http.get<ArrayBuffer>(url, { responseType: 'arraybuffer' })
     return Buffer.from(data)
   }
+
+  protected async analyze(url: string, file?: string): Promise<Assets.FileInfo> {
```
Comment (Shigma) [inline packages/koishi/src/assets.ts:34]: 这样是不是简单一些？

```ts
  protected async analyze(url: string, name = ''): Promise<Assets.FileInfo> {
    const buffer = await this.download(url)
    const hash = createHash('sha1').update(buffer).digest('hex')
    if (name) {
      name = basename(name)
      if (!name.startsWith('.')) {
        name = `-${name}`
      }
    } else {
      const fileType = await FileType.fromBuffer(buffer)
      if (fileType) name = `.${fileType.ext}`
    }
    return { buffer, hash, name, filename: `${hash}${name}` }
  }
```

Comment (purerosefallen) [reply packages/koishi/src/assets.ts:34]: exactly

**Review by Shigma — APPROVED** [review]
LGTM

**Pull request merged**